### PR TITLE
Add file count capability

### LINKED ISSUE #12: create new database tables
context: in order to monitor experiment run output files, we need the ability to save file count values into score-db. to achieve this, new tables are required to store applicable files for file types, storage locations, and the experiment file counts themselves. these tables will operate similarly to the existing experiment metrics tables. 

create new database tables for score-db to allow for saving of file counts

**Storage_location** -- id will be foreign key for other tables like experiment id
Template_path -- this would be key below
Platform
Bucket_name
-- there's a terminology for technology that goes cross-platform
Key - directory structure
Node - ?
Region (like east_2)

**File_types**
**Expt_stored_file_counts**

### LINKED ISSUE #13: refactor score-db to support file counts
update score-db to have calls to put and get data in the new file count tables following the existing infrastructure pattern

### PR DESCRIPTION
**Pull request by jrknezha**
This PR adds the ability to store and retrieve experiment file counts in the database. It also refactors some code from existing files into utilities that were found to be necessary for the new infrastructure as well to prevent further duplication of code for the new files. 

This PR will complete Milestone: Add ability to store file counts. Issues addressed by this PR include: #12 and #13. 

Improved features: 
Moved functions into the time_utils and created a db_utils and migrated some repeated calls into this new file from the existing table infrastructure (experiments, metric_types, etc.). 

New features: 
- table design to store file counts, storage locations, and experiment stored file counts. 
- the experiment stored file counts table is dependent on file counts, storage locations, and experiments 
- table interactions for file counts, storage locations, and experiment stored file counts for completing HTTP GET and PUT requests. this design is based on the existing infrastructure. 
- added new tables to the db request registry
- tests for the new table handlers 

Tests were run successfully and confirmed within the database that the new tables were created and the test items inserted successfully.

### CODE REVIEW

**Comment by frolovsa** [comment]
@jrknezha lets start by making an issue describing the problem we are trying to solve and adding to the description of this PR listing things that re changed/implemented so i can review this PR in more detail

**Comment by jrknezha** [comment]
@frolovsa I've added a milestone item to address the scope of the problem and connected the existing issues as well as updated the comment to give details of the updated and new features. It should be ready for your review now.

**Comment by frolovsa** [comment]
Thank you Jessica @jrknezha, I will take a look (at some point).

File: src/expt_metrics.py [hunk 1]
```diff
@@ -111,94 +90,52 @@ def __str__(self):
...
-        print(f'in post init name: {self.name}')
-        if self.cycle_start > self.cycle_stop:
-            msg = f'start time must be before end time - start: {self.cycle_start}, ' \
-                f'end: {self.cycle_stop}'
-            raise ValueError(msg)
-        if self.platform not in VALID_PLATFORMS:
-            msg = f'\'platform\' must be one of {VALID_PLATFORMS}, was ' \
-                f'\'{self.platform}\''
-            raise ValueError(msg)
+# @dataclass
```
Comment (frolovsa) [inline src/expt_metrics.py]: lets not commit large blocks of commented code.

File: src/expt_file_counts.py [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+    def __str__(self):
+        return self.message
+    
+@dataclass
+class ExptFileCount:
+    '''class for storing experiment file count values'''
+    count: int
+    folder_path: String
+    cycle: datetime
+    time_valid: datetime
```
Comment (frolovsa) [inline src/expt_file_counts.py]: I am not clear about the difference between cycle and time_valid. I guess cycle is like a directory (initial start date) and valid time is more like forecast length? If that is the case. should time_valid be forecast_length (real value in hours) instead?

File: src/file_types.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+    ],
+)
+
+@dataclass
+class FileType:
+    ''' file type object storing data related to the file info '''
+    name: str
+    file_template: str
+    file_format: str
+    stat_type: str
```
Comment (frolovsa) [inline src/file_types.py]: what would we store in stat_type?

File: src/file_types.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+        'description'
+    ],
+)
+
+@dataclass
+class FileType:
+    ''' file type object storing data related to the file info '''
+    name: str
+    file_template: str
+    file_format: str
```
Comment (frolovsa) [inline src/file_types.py:44]: Is there a reason for us to have a separate file_format column? (in addition to file_template?)

File: src/file_types.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+            ft.file_format,
+            ft.stat_type,
+            ft.description,
+            ft.created_at,
+            ft.updated_at
+        ).select_from(
+            ft
+        )
+
+        print('Before adding filters to file types request########################')
```
Comment (frolovsa) [inline src/file_types.py:278]: are we ok using print() as oppose to using logger class?

File: src/score_table_models.py [hunk 6]
```diff
@@ -168,11 +232,14 @@ def get_session():
 
 # def init_tables():
```
Comment (frolovsa) [inline src/score_table_models.py]: same comment about comitting commented code

File: tests/test_expt_file_counts_handler.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+def test_put_expt_file_counts_request_dict():
+    request_dict = {
+        'name': 'expt_file_counts',
+        'method': 'PUT',
+        'body': {
+            'experiment_name': 'C96L64.UFSRNR.GSI_3DVAR.012016',
+            'wallclock_start': '2021-07-22 09:22:05',
+            'file_type_name': 'example_type',
+            'file_extension': '.example',
+            'time_valid' : '2023-02-03 04:54:23',
```
Comment (frolovsa) [inline tests/test_expt_file_counts_handler.py]: note that time_valid here is smaller than the cycle value. This seem strange to me.

File: tests/test_expt_file_counts_handler.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            'storage_loc_name' : 's3_example_bucket',
+            'storage_loc_platform': 'aws_s3',
+            'storage_loc_key': 'noaa-exmample-score-db-bucket/reanalysis',
```
Comment (frolovsa) [inline tests/test_expt_file_counts_handler.py]: it stand out to me that `storage_loc_key` and `folder_path` have a lot of overlap. Is that a bug or a feature?

File: tests/test_storage_locations_handler.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+def test_storage_location_input_dict():
+    request_dict = {
+        'name': 'storage_locations',
+        'method': 'PUT',
+        'body': {
+            'name': 's3_example_bucket',
+            'platform': 'aws_s3', 
+            'bucket_name': 'noaa-example-score-db-bucket',
+            'key': 'noaa-exmample-score-db-bucket/reanalysis',
```
Comment (frolovsa) [inline tests/test_storage_locations_handler.py]: should the key really have a bucket_name in it? Seems like it will not in a real world?

File: src/expt_metrics.py (hunk 1 above)
Comment (frolovsa) [reply src/expt_metrics.py]: also, where did ExptMetric class go? Was it removed by accident or on purpose?

**Review by frolovsa — CHANGES_REQUESTED** [review]
Thank you Jessica @jrknezha for doing all this work. I didn't realize the amount of the new code that was added. I think this is fine overall. I find that the code is clean. I have a few clean up comments and a few comments about the structure of the columns in the new tables. My comments about the new tables might require a conversation. 

One bigger comment is that we probably should also update the README file to reflect the ew tables. 

Thank you again for this contribution and working through the issues that came up.

File: src/expt_metrics.py (hunk 1 above)
Comment (jrknezha) [reply src/expt_metrics.py]: so I commented it out to prove it wasn't used in the workflow when running tests because I couldn't find any reference to it in the rest of the code; I should have fully deleted it because it's not needed.

File: src/file_types.py (hunk 4 above)
Comment (jrknezha) [reply src/file_types.py:44]: I was envisioning file template as a reference to the name of the file (i.e. ocean-%Y%m.nc) but the format would be netcdf, text, CSV, or something along those lines. I can see that being relevant in future in regards to harvesters designed for certain file formats. Or if it's a customized file like the text logs are output in csv.

File: src/file_types.py (hunk 3 above)
Comment (jrknezha) [reply src/file_types.py]: I was thinking potentially something along 'innovation stats' or 'ocean' or 'atmosphere' but that may not be needed since our discussion about the overall file type being ocean or atmosphere

File: src/file_types.py (hunk 5 above)
Comment (jrknezha) [reply src/file_types.py:278]: I think it's better for now to use print() since the entire repo is designed that way and it's useful in the cylc suite output. Perhaps in the future a logger would be better going forward but it would require editing most, if not all, the files in src.

File: tests/test_expt_file_counts_handler.py (hunk 7 above)
Comment (jrknezha) [reply tests/test_expt_file_counts_handler.py]: I understood time valid as the time in which the value referred to was accurate so the current time the metric was calculated in case those files change in a new run of an experiment. I was following the pattern also used for experiment metrics. How did you understand time valid being used? I think we should have a further discussion about the table schema and column usage before we finalize it for use in production

File: tests/test_expt_file_counts_handler.py (hunk 8 above)
Comment (jrknezha) [reply tests/test_expt_file_counts_handler.py]: for me it's a feature in that the key is the top level and the folder path was the full path to the specific file being counted. the key is a value that is stored in the storage locations table and used to identify which storage location the file count is associated with whereas the folder path is only in relation to this specific count object.

File: src/expt_metrics.py (hunk 1 above)
Comment (frolovsa) [reply src/expt_metrics.py]: Sounds good.

File: src/file_types.py (hunk 3 above)
Comment (frolovsa) [reply src/file_types.py]: yeh. `name` might be good enough.

File: src/file_types.py (hunk 4 above)
Comment (frolovsa) [reply src/file_types.py:44]: OK. i just couldn't see a use case for a search under `file_format`. It's up to you.

File: src/file_types.py (hunk 5 above)
Comment (frolovsa) [reply src/file_types.py:278]: got it. I didnt realize that. Resolving this comment.

File: tests/test_expt_file_counts_handler.py (hunk 7 above)
Comment (frolovsa) [reply tests/test_expt_file_counts_handler.py]: Let chat about this one in person on tuesday.

File: tests/test_expt_file_counts_handler.py (hunk 8 above)
Comment (frolovsa) [reply tests/test_expt_file_counts_handler.py]: ok. boto3 (aws python lib) usually requires bucket and key to be separate. E.g. technically, for aws, `noaa-exmample-score-db-bucket` is not part of the key.

File: src/db_utils.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+def validate_column_name(cls, value):
+    if not isinstance(value, str):
+        raise TypeError(f'Column name must be a str, was {type(value)}')
+
+    try:
+        column_obj = getattr(cls, value)
+        print(f'column: {column_obj}, type(key): {type(column_obj)}')
+    except Exception as err:
+        msg = f'Column does not exist - err: {err}'
+        raise ValueError(msg)
```
Comment (amschne) [inline src/db_utils.py:63]: I'm seeing a lot of these except statements (i.e., "except Exception as err:") throughout score-db/hv. What is the rational for this usage? Could, instead, there be a more specific Error type that we want to catch, so that the custom error message and corresponding ValueError (in this case) is raised only when the expected error occurs? In other words, what is the expected error type that is raised if the "column does not exist"? Could the except statement be replaced with "except [SPECIFIC ERROR TYPE] as err"?

File: src/db_utils.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    
+    return column_obj
+
+
+def validate_order_dir(value):
+    if not isinstance(value, str):
+        raise TypeError(f'\'order_by\' must be a str, was {type(value)}')
+    
+    if value not in VALID_ORDER_BY:
+        raise TypeError(f'\'order_by\' must be one of {VALID_ORDER_BY}, ' \
```
Comment (amschne) [inline src/db_utils.py]: Is the back-slash needed here? I think not because the line continuation is within parenthesis.

File: src/db_utils.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    ordering: list - this is a list of dicts which describe all the desired
+        column data sequential ordering (or the ORDER BY sql clause).
+
+        example list of orderby dicts: 
+        [
+            {'name': 'group_id', 'order_by': 'desc'},
+            {'name': 'created_at', 'order_by': 'desc'}
+        ]
+
+    """
```
Comment (amschne) [inline src/db_utils.py:98]: Nicely formatted docstring

File: src/db_utils.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    if ordering is None:
+        return None
```
Comment (amschne) [inline src/db_utils.py:100]: Because of the swift "return None" here, is this entire function optional to the broader program? i.e., in the higher level scope, what happens if this function returns None without erroring out?

File: src/experiments.py [hunk 14]
```diff
@@ -22,6 +22,8 @@
 from db_action_response import DbActionResponse
 import score_table_models as stm
 from score_table_models import Experiment as exp
+import time_utils
+import db_utils
```
Comment (amschne) [inline src/experiments.py:26]: yes! much appreciated refactoring of these utilities

File: src/expt_file_counts.py [hunk 15]
```diff
@@ -0,0 +1,646 @@
...
+        'time_valid',
+        'forecast_length',
+        'experiment_id',
+        'file_type_id',
+        'storage_location_id',
+    ]
+)
+
+class ExptFileCountsError(Exception):
+    def __init__(self, m):
```
Comment (amschne) [inline src/expt_file_counts.py]: change "m" to "message", for clarity? would avoid single character variable

File: src/expt_file_counts.py [hunk 16]
```diff
@@ -0,0 +1,646 @@
...
+    
+    def __repr__(self):
+        return f'expt_file_count_data: {self.expt_file_count_data}'
+    
+    def get_expt_file_count_data(self):
+        return self.expt_file_count_data
+    
+def get_file_count_from_body(body):
+    if not isinstance(body, dict):
+        msg = 'The \'body\' key must be a type dict, actually ' \
```
Comment (amschne) [inline src/expt_file_counts.py]: I noticed before that these types of custom error messages have "was" in place of "actually". Just noting that in case there is a desire for consistency.

File: src/expt_file_counts.py [hunk 17]
```diff
@@ -0,0 +1,646 @@
...
+    value = filter_dict.get(key)
+    if value is None:
+        print(f'No \'{key}\' filter detected')
+        return constructed_filter
+
+    exact_datetime = time_utils.get_time(value.get(db_utils.EXACT_DATETIME))
+
+    if exact_datetime is not None:
+        constructed_filter[key] = (
+            getattr(cls, key) == exact_datetime
```
Comment (amschne) [inline src/expt_file_counts.py:140]: This looks like the intent is to return constructed_filter, a dictionary of True or False values here. Is that correct?

File: src/expt_file_counts.py [hunk 18]
```diff
@@ -0,0 +1,646 @@
...
+    return constructed_filter
+
+def validate_list_of_strings(values):
+    if isinstance(values, str):
+        val_list = []
+        val_list.append(values)
+        return val_list
+
+    if not isinstance(values, list):
+        msg = f'string values must be a list - actually: {type(values)}'
```
Comment (amschne) [inline src/expt_file_counts.py]: again, purely for consistency, this type of custom error message before used a comma, not a dash, to separate the "actually..." or "was... " clause

File: src/expt_file_counts.py [hunk 19]
```diff
@@ -0,0 +1,646 @@
...
+            'record_limit': 1
+        }
+    }
+    print(f'expt_request: {expt_request}')
+    er = ExperimentRequest(expt_request)
+    results = er.submit()
+    print(f'results: {results}')
+
+    record_cnt = 0
+    try:
```
Comment (amschne) [inline src/expt_file_counts.py:291]: this try/except block, with the multiple if statements, reads a little enigmatic (to me at least). could the order of try versus if statements instead be reversed, so that there are multiple small (tiny) try/except blocks, each within a corresponding if statement?. Breaking down this try block in this manner could improve the readability plus refine the custom error catching / messaging.

File: src/expt_file_counts.py [hunk 20]
```diff
@@ -0,0 +1,646 @@
...
+        error_msg = f'Problem finding storage location id from request: {storage_loc_request} '\
+            f'- err: {err}'
+        print(f'error_msg: {error_msg}')
+        raise ExptFileCountsError(error_msg)
+
+    return storage_loc_id
+
+
+@dataclass
+class ExptFileCountRequest:
```
Comment (amschne) [inline src/expt_file_counts.py:417]: since this is a longer class, consider adding a docstring

File: src/expt_file_counts.py [hunk 21]
```diff
@@ -0,0 +1,646 @@
...
+        )
+
+        constructed_filter = get_time_filter(filters, esfc, 'cycle', constructed_filter)
+
+        if len(constructed_filter) > 0:
+            try: 
+                for key,value in constructed_filter.items():
+                    print(f'adding filter: {value}')
+                    query = query.filter(value)
+            except Exception as err:
```
Comment (amschne) [inline src/expt_file_counts.py:515]: I think this exception might produce obscure error messages. Could "Problems" be made more specific?

File: src/storage_locations.py [hunk 22]
```diff
@@ -0,0 +1,328 @@
...
+    storage_location: StorageLocation = field(init=False)
+    storage_location_data: namedtuple = field(init=False)
+    response: dict = field(default_factory=dict, init=False)
+
+    def __post_init__(self):
+        self.method = db_utils.validate_method(self.request_dict.get('method'))
+        self.params = self.request_dict.get('params')
+
+        self.body = self.request_dict.get('body')
+        if self.method == db_utils.HTTP_PUT:
```
Comment (amschne) [inline src/storage_locations.py:154]: just noting here that there are 3 levels of nested if statements - not sure of a better workaround at the moment

**Review by amschne — COMMENTED** [review]
Wow, Jessica, these additions are a heavy lift! Thank you for all this work. I can't wait to start testing.

It's clear that you're diligently working to revive this repo to the point of having something running, so I don't think my concerns are critical at the moment. But here are my initial comments:

1. Generally, the use of longer try blocks with a lot going on within them worries me. There are also quite a few "except Exception as err:" statements (which I realize may have been inherited form previous versions) that serve as a "catch all" for anything that might go wrong within. This combined with the more complex try blocks could result in a difficult debugging process for a new user. I encountered this problem in socre-hv, and selfishly would like to preempt this confusing, possibly time-consuming debugging process in the future. For this reason, I suggest using very short try blocks. The existing custom error messages are (mostly, with a couple outliers) specific enough to detect a problem, given that it is the anticipated issue and not something else caught. To avoid the latter, I suggest using more specific Error types in the except statements.

2. I really appreciate the refactoring of utilities. This will undoubtably prevent headaches if and when constants need to be changed in the future.

3. Docstring and in line comments are sparse.

My in line comments are mostly minor. For example, I noted some inconsistency in the custom error messages, which could be addressed for cosmetic purposes. Some logic reads a little confusing, which could be improved with in line comments. I saw one triple nested if block that leads me to think about multiple corner cases that would be tedious to exercise in writing unit tests.

I realize this PR is a first stab at getting this repo working, so please feel free to address my comments as you see fit. Thanks again!

File: tests/test_expt_file_counts_handler.py (hunk 7 above)
Comment (jrknezha) [reply tests/test_expt_file_counts_handler.py]: updated based on in person discussion

**Comment by jrknezha** [comment]
At this point, the new branch has been run successfully for the new tables and works in conjunction with score-monitoring.

File: src/db_utils.py (hunk 13 above)
Comment (jrknezha) [reply src/db_utils.py:100]: Yes, this is optional, it is dependent on the inputs from the user (or other calls from within the program). ordering is not required to make a call to the database so None will not break it.

File: src/db_utils.py (hunk 12 above)
Comment (jrknezha) [reply src/db_utils.py:98]: I won't take credit here, this is from Steve's work I just moved it to a new file so it looks new. Good practice to strive for!

File: src/expt_file_counts.py (hunk 17 above)
Comment (jrknezha) [reply src/expt_file_counts.py:140]: The dictionary is of keys and sqlalchemy structures

File: src/db_utils.py (hunk 10 above)
Comment (jrknezha) [reply src/db_utils.py:63]: As discussed in person, this would be a much larger change then intended for this PR for the structure of the repo and as written, currently is effective for debugging. This is a change to consider in future as part of a larger rewrite of the repo if desired.

File: src/db_utils.py (hunk 11 above)
Comment (jrknezha) [reply src/db_utils.py]: addressed

File: src/expt_file_counts.py (hunk 16 above)
Comment (jrknezha) [reply src/expt_file_counts.py]: cleaned up for consistency, good catch

File: src/expt_file_counts.py (hunk 18 above)
Comment (jrknezha) [reply src/expt_file_counts.py]: cleaned up for consistency

File: src/expt_file_counts.py (hunk 19 above)
Comment (jrknezha) [reply src/expt_file_counts.py:291]: as discussed in person, this would be a larger rewrite and most of the code block could not be separated into smaller catches for this logic. something to be considered for a future clean up of the repo.

File: src/expt_file_counts.py (hunk 21 above)
Comment (jrknezha) [reply src/expt_file_counts.py:515]: not with my current understanding of the repo, however, this code block does return the full error message which is useful in determining the issue

File: src/expt_file_counts.py (hunk 15 above)
Comment (jrknezha) [reply src/expt_file_counts.py]: done

File: src/expt_file_counts.py (hunk 20 above)
Comment (jrknezha) [reply src/expt_file_counts.py:417]: added

**Review by frolovsa — APPROVED** [review]
@jrknezha this is indeed a heavy lift. I haven't re-reviewed. just going ahead and approve it. i saw that you and adam have worked on this one together.
